webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx

```diff
@@ -576,6 +609,98 @@ const msg = defineMessages({
     fileInputEdit: {id: 'channel_bookmarks.create.file_input.edit', defaultMessage: 'Edit'},
-    linkInvalid: {id: 'channel_bookmarks.create.error.invalid_url', defaultMessage: 'Please enter a valid link'},
+    linkInvalid: {id: 'channel_bookmarks.create.error.invalid_url', defaultMessage: 'Please enter a valid link. Could not parse: {link}.'},
     saveError: {id: 'channel_bookmarks.create.error.generic_save', defaultMessage: 'There was an error trying to save the bookmark.'},
 });
+
+const TYPING_DELAY_MS = 250;
+const REQUEST_TIMEOUT = 10000;
+export const useBookmarkLinkValidation = (link: string, onValidated: (validatedLink: string, forced?: boolean) => void) => {
+    const {formatMessage} = useIntl();
+
+    const [loading, setLoading] = useState<URL>();
+    const [error, setError] = useState<ReactNode>();
+    const [suppressed, setSuppressed] = useState(false);
+    const abort = useRef<AbortController>();
+
+    const start = useCallback((url: URL) => {
+        setLoading(url);
+        setError(undefined);
+        setSuppressed(false);
+        abort.current = new AbortController();
+        return abort.current.signal;
+    }, []);
+
+    const cancel = useCallback(() => {
+        abort.current?.abort('stale request');
+        abort.current = undefined;
+        setLoading(undefined);
+    }, []);
+
+    useEffect(() => {
+        const handler = setTimeout(async () => {
+            cancel();
+            if (!link) {
+                return;
+            }
+
+            const url = validHttpUrl(link);
+
+            if (!url) {
+                setError(formatMessage(msg.linkInvalid, {link}));
+                setLoading(undefined);
+                return;
+            }
+
+            const signal = start(url);
+
+            try {
+                await fetch(url, {
+                    mode: 'no-cors',
+
+                    // @ts-expect-error AbortSignal.any exists; remove this directive when next TS upgrade
+                    signal: AbortSignal.any([signal, AbortSignal.timeout(REQUEST_TIMEOUT)]),
+                });
+                onValidated(link);
+            } catch (err) {
+                if (signal === abort.current?.signal) {
+                    setError(continuableLinkErr(url, () => {
+                        onValidated(link, true);
+                        setSuppressed(true);
+                        setError(continuableLinkErr(url));
+                    }));
+                }
+            } finally {
+                setLoading((currentUrl) => {
+                    if (currentUrl !== url) {
+                        // trailing effect of cancelled
+                        return currentUrl;
+                    }
+
+                    return undefined;
+                });
+            }
+        }, TYPING_DELAY_MS);
+
+        return () => clearTimeout(handler);
+    }, [link, start, cancel]);
+
+    return [error, {loading: Boolean(loading), suppressed}] as const;
```

Comment:
Yes we could, and I'm not strongly opposed to that idea, but 2/5 or 3/5 I have come to like this return pattern for custom hooks.

```tsx
return [result, extraContextAndOrControls] as const
``` 

It mimics the array type of `useState`, `useDispatch`, `useTransition`,  etc. in that the first item is the subject/result or main return value, and the second is the setter, or in these cases—extra context, status and controls or actions that may or may not be needed by the caller.

Reply:
I think that pattern is more helpful for more general hooks like `useState` and `useTransition` where they may be called multiple times and we need to rename the returned values to reflect their meaning, but I don't know if I like it as much here. Destructuring a tuple or an object is almost identical, but I've never liked destructuring nested objects since it's never felt like an intuitive pattern to me (especially if that ever involves renaming things since I still forget the order you put the new and old names in).

That all said, I'm only 1/5 on that though since both behave the same way, and thanks to TS and intellisense, we don't need to even really think about that when using it